pkg/controller/issueapi/handle_issue.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package issueapi
+
+import (
+	"errors"
+	"net/http"
+
+	"github.com/google/exposure-notifications-verification-server/pkg/api"
+	"github.com/google/exposure-notifications-verification-server/pkg/controller"
+	"github.com/google/exposure-notifications-verification-server/pkg/observability"
+)
+
+// HandleIssue responds to the /issue API for issuing verification codes
+func (c *Controller) HandleIssue() http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if result := c.handleIssueFn(w, r); result != nil {
+			recordObservability(r.Context(), result)
+		}
+	})
+}
+
+func (c *Controller) handleIssueFn(w http.ResponseWriter, r *http.Request) *issueResult {
```

Comment:
I'm also not sure I see the value in pulling this out into its own function as-written. It just makes it harder to test because it's package private, but it's the same signature as the public function.

I was hoping we could do something more like:

```go
func (c *Controller) HandleIssueAPI() http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
    ctx := r.Context()

    realm := controller.RealmFromContext(ctx)
    if realm == nil {
      controller.MissingRealm(w, r, c.h)
      return
    }

    authApp := controller.AuthorizedAppFromContext(ctx)
    if authApp == nil {
      controller.MissingAuthorizedApp(w, r, c.h)
      return
    }

    if authApp.Type != database.APIKeyTypeAdmin {
      controller.Unauthorized(w, r, c.h)
      return
    }

    if err := c.handleIssue(ctx, realm, authApp, nil); err != nil {
      // build and return response
    }
	})
}

func (c *Controller) HandleIssueUI() http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
    ctx := r.Context()

    membership := controller.MembershipFromContext(ctx)
    if membership == nil {
      controller.MissingMembership(w, r, c.h)
      return
    }
    realm := membership.Realm

    if !membership.Can(rbac.CodeIssue) {
      controller.Unauthorized(w, r, c.h)
      return
    }

    if err := c.handleIssue(ctx, realm, nil, membership); err != nil {
      // build and return response
    }
	})
}

func (c *Controller) handleIssue(ctx context.Context, realm *database.Realm, authApp *database.AuthorizedApp, membership *database.Membership) *issueResult {
  // ...
}
```

Then you've separated rbac handling and http stuff out of the core logic.

Reply:
it pulls out http.HandlerFunc so I can test without real mux routing.
The real reason I did this was to `recordObservability` here because the variable capture of `defer` was confusing me

Reply:
But I like your idea too, makes it less weird to be serving this at two routes